### PR TITLE
[webui] add titles/popovers for build states

### LINKED ISSUE #672: Titles/Popovers for build states
There are a lot of places where we show the build status of packages, but the documentation is only on project_controller.rb#monitor. This often leads to confusion.

### PR DESCRIPTION
**Pull request by mschnitzer**
This fixes #672

Here are a few screenshots:
preview-1

(Hovering over the info icon will display a tooltip with the explanation of the current status code)

screenshot-6

screenshot-7

### CODE REVIEW

File: src/api/app/helpers/webui/buildstatus_helper.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
+module Webui::BuildstatusHelper
+  # rubocop:disable Style/AlignHash
+  PACKAGE_STATUS_LIST = {
+      'succeeded'     => 'Package has built successfully and can be used to build further packages.',
+      'failed'        => 'The package does not build successfully. No packages have been created. Packages ' \
```
Comment (bgeuken) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: Please use '+' here instead of a backslash

File: src/api/app/helpers/webui/buildstatus_helper.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      'unresolvable'  => 'The build can not begin, because required packages are either missing or not explicitly defined.',
+      'broken'        => 'The sources either contain no build description (e.g. specfile), automatic source processing failed or a ' \
+                       'merge conflict does exist.',
+      'blocked'       => 'This package waits for other packages to be built. These can be in the same or other projects.',
+      'scheduled'     => 'A package has been marked for building, but the build has not started yet.',
+      'dispatching'   => 'A package is being copied to a build host. This is an intermediate state before building.',
+      'building'      => 'The package is currently being built.',
+      'signing'       => 'The package has been built successfully and is assigned to get signed.',
+      'finished'      => 'The package has been built and signed, but has not yet been picked up by the scheduler. This is an ' \
+                       'intermediate state prior to \'succeeded\' or \'failed\'.',
```
Comment (bgeuken) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: Use double quotes when you have single quotes in your string. Or even better default to double quotes:-)

File: src/api/app/helpers/webui/buildstatus_helper.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
+module Webui::BuildstatusHelper
+  # rubocop:disable Style/AlignHash
```
Comment (bgeuken) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: Why is this needed? The hash looks well aligned to me.

File: src/api/app/helpers/webui/buildstatus_helper.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      'dispatching'   => 'A package is being copied to a build host. This is an intermediate state before building.',
+      'building'      => 'The package is currently being built.',
+      'signing'       => 'The package has been built successfully and is assigned to get signed.',
+      'finished'      => 'The package has been built and signed, but has not yet been picked up by the scheduler. This is an ' \
+                       'intermediate state prior to \'succeeded\' or \'failed\'.',
+      'disabled'      => 'The package has been disabled from building in project or package metadata.',
+      'excluded'      => 'The package build has been disabled in package build description (for example in the .spec file) or ' \
+                       'does not provide a matching build description for the target.',
+      'unknown'       => 'The scheduler has not yet evaluated this package. Should be a short intermediate state for new packages.'
+  }.with_indifferent_access
```
Comment (bgeuken) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: Given that you use with_indifferent_access you can, and should, convert the hash keys to symbol format -> { disabled: "foo" }

File: src/api/app/helpers/webui/buildstatus_helper.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  # rubocop:enable Style/AlignHash
+
+  def get_package_status_description(status)
+    PACKAGE_STATUS_LIST[status] || "status explanation not found"
+  end
+
+  def arch_repo_table_cell(repo, arch, package_name)
+    status = @statushash[repo][arch][package_name] || { 'package' => package_name }
```
Comment (bgeuken) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: Please use { package: package_name }, unless this breaks something

Comment (bgeuken) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: Looks like it's expected to be a stringified key though

File: src/api/app/helpers/webui/buildstatus_helper.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    end
+
+    result = "<td class='".html_safe
+    result += "#{theclass}"
+    result +=" buildstatus nowrap'>".html_safe
+
+    if %w(- unresolvable blocked excluded scheduled).include?(code)
+      result += link_to(code, '#', title: link_title, id: status_id, class: code)
+    else
+      result += link_to(code.gsub(/\s/, '&nbsp;'),
```
Comment (bgeuken) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: Depending on where the code variable comes from (user input?), we should escape it. Or mark it as unsafe.

File: src/api/app/helpers/webui/buildstatus_helper.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+                        {
+                            action: :live_build_log, package: package_name, project: @project.to_s,
+                            arch: arch, controller: 'package', repository: repo
+                        },
+                        { title: link_title, rel: 'nofollow' }
+                       )
+    end
+
+    if !status['code'].nil?
```
Comment (bgeuken) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: ``` if status['code'] ```

File: src/api/app/helpers/webui/buildstatus_helper.rb [hunk 8]
```diff
@@ -53,7 +53,7 @@ def arch_repo_table_cell(repo, arch, package_name)
                        )
     end
 
-    if !status['code'].nil?
+    if status['code']
```
Comment (bgeuken) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: Please rebase this change with the previous commit.

**Comment by bgeuken** [comment]
Regarding the commit explanation above: "This fixes some issues with the icons in the webui."
It would be good to mention issues that get solved by this commit:-)

File: src/api/app/helpers/webui/buildstatus_helper.rb (hunk 2 above)
Comment (mschnitzer) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: OK, I was a bit confused what to use by default since we use both across files.

File: src/api/app/helpers/webui/buildstatus_helper.rb (hunk 3 above)
Comment (mschnitzer) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: Rubocop complains about the indentation. I talked to several developers here and no one could really help me. The only solution what I found was "disable it". It seems this is a bug in rubocop.

Comment (bgeuken) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: Did you try to align the lines you moved to a newline with the indentation level of the other hash values? eg. line 6, 9, 16

**Comment by mschnitzer** [comment]
I changed the title to make it more clear.

File: src/api/app/helpers/webui/buildstatus_helper.rb (hunk 6 above)
Comment (ChrisBr) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: It should be unsafe because result is an ActiveSupport::SafeBuffer, concatenation of a String and an ActiveSupport::SafeBuffer will produce a new AtiveSupport::SafeBuffer

File: src/api/app/helpers/webui/buildstatus_helper.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+                            action: :live_build_log, package: package_name, project: @project.to_s,
+                            arch: arch, controller: 'package', repository: repo
+                        },
+                        { title: link_title, rel: 'nofollow' }
+                       )
+    end
+
+    if status['code']
+      status_desc = get_package_status_description(status['code'])
+      result += " #{sprite_tag 'help', title: status_desc}".html_safe
```
Comment (ChrisBr) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: I'd split this in two lines because it is hard to read

Comment (ChrisBr) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: And then you also don't need the temp variable ``status_desc``

File: src/api/app/helpers/webui/buildstatus_helper.rb [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+                       "does not provide a matching build description for the target.",
+      unknown:       "The scheduler has not yet evaluated this package. Should be a short intermediate state for new packages."
+  }.with_indifferent_access
+  # rubocop:enable Style/AlignHash
+
+  def get_package_status_description(status)
+    PACKAGE_STATUS_LIST[status] || "status explanation not found"
+  end
+
+  def arch_repo_table_cell(repo, arch, package_name)
```
Comment (ChrisBr) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: What about removing ``arch_repo_table_cell`` function from ``src/api/app/helpers/webui/webui_helper.rb``?

File: src/api/app/helpers/webui/buildstatus_helper.rb [hunk 11]
```diff
@@ -0,0 +1,63 @@
+module Webui::BuildstatusHelper
+  # rubocop:disable Style/AlignHash
+  PACKAGE_STATUS_LIST = {
```
Comment (ChrisBr) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: Sure that this should live in a helper? For me it looks like Buildresult or Package model is more appropriate?

File: src/api/app/helpers/webui/webui_helper.rb [hunk 12]
```diff
@@ -423,4 +423,9 @@ def replace_jquery_meta_characters(input)
     # The stated characters are c&p from https://api.jquery.com/category/selectors/
     input.gsub(/[!"#$%&'()*+,.\/:\\;<=>?@\[\]^`{|}~]/, '_')
   end
+
+  def word_break(string, length)
+    # adds a <wbr> tag after an amount of given characters
+    string.scan(/.{1,#{length}}/).join("<wbr>")
```
Comment (ChrisBr) [inline src/api/app/helpers/webui/webui_helper.rb]: Does that work? I would assume that you need to mark ``.join("<wbr>")`` as ``html_safe``

File: src/api/app/helpers/webui/buildstatus_helper.rb (hunk 3 above)
Comment (bgeuken) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: @mschnitzer Between unresolvable: and it's value you have 2 whitespaces. When you remove that one, it passes in rubcop.

File: src/api/app/helpers/webui/webui_helper.rb (hunk 12 above)
Comment (ChrisBr) [inline src/api/app/helpers/webui/webui_helper.rb]: (but just an assumption, not tried ...)

File: src/api/app/helpers/webui/buildstatus_helper.rb (hunk 10 above)
Comment (bgeuken) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: @mschnitzer IIRC some of your commits got reverted last week. Might be that the other method slipped back in. Please check

Comment (ChrisBr) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: Yes, @adrianschroeter reverted some of your commits before deploying last week.

Comment (mschnitzer) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: I thought I removed the method form the webui_helper already. Anyway, I did that right now.

File: src/api/app/views/webui/project/_buildstatus.html.erb [hunk 13]
```diff
@@ -10,25 +10,29 @@
...
         <% @buildresult.each do |repo, archarray| %>
             <% index = 0
-               srepo = truncate(repo, :length => 17)
                archarray.each do |arch, counts| %>
                 <tr>
                   <% if index == 0 %>
-                      <td class="repo nowrap" rowspan="<%= archarray.length %>">
-                        <%= link_to(elide(srepo, 26), { action: :state, project: @project.name, repository: repo, controller: :repositories }, { title: "Repository #{srepo}"}) %>
+                      <td class="repo" rowspan="<%= archarray.length %>">
+                        <%= link_to(word_break(repo, 13), { action: :state, project: @project.name, repository: repo, controller: :repositories }, { title: "Repository #{repo}"}) %>
```
Comment (ChrisBr) [inline src/api/app/views/webui/project/_buildstatus.html.erb]: Why 13? Looks like a "magic number"

Comment (ChrisBr) [inline src/api/app/views/webui/project/_buildstatus.html.erb]: And I would prefer repository as variable name

File: src/api/app/views/webui/package/_buildstatus.html.erb [hunk 14]
```diff
@@ -12,8 +12,8 @@
           -%>
           <tr>
             <% if index == 0 %>
-              <td class="nowrap" title="<%= r %>" rowspan="<%= archs.length %>">
-                <%= link_to(elide(r, 20), {:action => :binaries, :controller => :package, :project => @project, :package => @package, :repository => r}, {:title => "Binaries for #{r}"}) %>
+              <td title="<%= r %>" rowspan="<%= archs.length %>">
+                <%= link_to(word_break(r, 22), {:action => :binaries, :controller => :package, :project => @project, :package => @package, :repository => r}, {:title => "Binaries for #{r}"}) %>
```
Comment (ChrisBr) [inline src/api/app/views/webui/package/_buildstatus.html.erb]: Same for 22

File: src/api/app/views/webui/project/_buildstatus.html.erb (hunk 13 above)
Comment (mschnitzer) [inline src/api/app/views/webui/project/_buildstatus.html.erb]: `13` is the amount of characters what fit in a column.

File: src/api/app/views/webui/package/_buildstatus.html.erb (hunk 14 above)
Comment (mschnitzer) [inline src/api/app/views/webui/package/_buildstatus.html.erb]: Same answer :-)

Comment (ChrisBr) [inline src/api/app/views/webui/package/_buildstatus.html.erb]: Which is?

Comment (ChrisBr) [inline src/api/app/views/webui/package/_buildstatus.html.erb]: copy from mail as the comment vanished:

> 13 is the amount of characters what fit in a column.

For every size of screen?
However, it makes it not more readable nevertheless ...

File: src/api/app/helpers/webui/buildstatus_helper.rb (hunk 6 above)
Comment (mschnitzer) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: So, what should we dor here? I didn't touch the code, I just moved it to this file.

File: src/api/spec/helpers/webui/webui_helper_spec.rb [hunk 15]
```diff
@@ -40,6 +40,21 @@
...
   end
 
+  describe '#word_break' do
+    it "continuously adds tag after N characters" do
+      expect(word_break("XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX", 10)).to \
+        eq("XXXXXXXXXX<wbr>XXXXXXXXXX<wbr>XXXXXXXXXX<wbr>XXXXXXXXXX")
+    end
+
+    it "adds no tag if string is shorter than N characters" do
+      expect(word_break("XX", 10)).to eq("XX")
```
Comment (ChrisBr) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: I would test 10 characters as this is the boundary, not?

File: src/api/app/helpers/webui/buildstatus_helper.rb (hunk 6 above)
Comment (ChrisBr) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: I think it is fine...

File: src/api/app/views/webui/package/_buildstatus.html.erb (hunk 14 above)
Comment (ChrisBr) [inline src/api/app/views/webui/package/_buildstatus.html.erb]: As it was the same before it is fine for me...

File: src/api/spec/helpers/webui/webui_helper_spec.rb (hunk 15 above)
Comment (mschnitzer) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: Not really sure if I got you right. Since the test tests if NO tag was added if the given amount of characters is shorter than the second parameter.

File: src/api/app/views/webui/package/_buildstatus.html.erb (hunk 14 above)
Comment (mschnitzer) [inline src/api/app/views/webui/package/_buildstatus.html.erb]: OK

File: src/api/spec/helpers/webui/webui_helper_spec.rb (hunk 15 above)
Comment (ChrisBr) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: Yes, right! But the second parameter is in this case 10, so either you change it to 2 and test that still no tag has applied or you leave the parameter to 10 but add 10 characters as first parameter. 

And your assumption 
> NO tag was added if the given amount of characters is shorter than 
is wrong! Because it also doesn't add an character if the input string length == parameter!

Comment (ChrisBr) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: So what I want to say: You should test the boundaries!

And as a suggestion to improve readability: I would change `x...x` to `1234..789` as I don't have to manually count the characters!

File: src/api/app/helpers/webui/webui_helper.rb [hunk 16]
```diff
@@ -423,4 +392,9 @@ def replace_jquery_meta_characters(input)
     # The stated characters are c&p from https://api.jquery.com/category/selectors/
     input.gsub(/[!"#$%&'()*+,.\/:\\;<=>?@\[\]^`{|}~]/, '_')
   end
+
+  def word_break(string, length)
+    # adds a <wbr> tag after an amount of given characters
+    string.scan(/.{1,#{length}}/).join("<wbr>").html_safe
```
Comment (ChrisBr) [inline src/api/app/helpers/webui/webui_helper.rb]: ``
string.scan(/.{1,#{length}}/).join("<wbr>".html_safe)
``
And I would test that it does escape if you insert html!

Comment (bgeuken) [inline src/api/app/helpers/webui/webui_helper.rb]: Adding a ``` string.expects(:html_safe).once ``` in the test should be good enough. What do you say @ChrisBr ?

Comment (bgeuken) [inline src/api/app/helpers/webui/webui_helper.rb]: On the other hand that wouldn't garanty that html_safe was could at the end of the process... nevermind

Comment (bgeuken) [inline src/api/app/helpers/webui/webui_helper.rb]: But we could test that the resulting string has tagged as safe. afaik it's inheriting from ActiveSupport::SafeBuffer then

**Comment by hennevogel** [comment]
Please include screenshots of how this looks like :-)

**Comment by mschnitzer** [comment]
@hennevogel sure, I'm still testing something regarding the new `rubocop` version (it complains now about `html_safe`) and then I'm gonna squash everything and add screenshots to it ;-)

**Comment by mschnitzer** [comment]
@bgeuken @hennevogel @ChrisBr  Could someone of you have a look again, please?

File: src/api/app/helpers/webui/buildstatus_helper.rb [hunk 17]
```diff
@@ -0,0 +1,59 @@
+module Webui::BuildstatusHelper
+  PACKAGE_STATUS_LIST = {
```
Comment (ChrisBr) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: Thought we wanted to move the hash into the model?

@bgeuken ^^

File: src/api/app/helpers/webui/webui_helper.rb (hunk 16 above)
Comment (ChrisBr) [inline src/api/app/helpers/webui/webui_helper.rb]: I still think that you should only apply the ``html_safe`` to the ``"<wbr>"``

File: src/api/app/helpers/webui/buildstatus_helper.rb (hunk 17 above)
Comment (mschnitzer) [inline src/api/app/helpers/webui/buildstatus_helper.rb]: Oh, I forgot to do that. This PR has way too much comments haha :-D

File: src/api/app/models/buildresult.rb [hunk 18]
```diff
@@ -19,6 +19,30 @@ class Buildresult < ActiveXML::Node
   }
   # rubocop:enable Style/AlignHash
 
+  STATUS_DESCRIPTION_LIST = {
```
Comment (ChrisBr) [inline src/api/app/models/buildresult.rb]: STATUS_DESCRIPTION

**Comment by bgeuken** [comment]
Not sure what you agreed on with Henne, but I would like to let the tooltip also appear when one hovers the mouse over the 'succeed', 'error', etc text. @openSUSE/open-build-service What do you say?

**Comment by bgeuken** [comment]
There still seem to be spacing issues on the monitoring page. If I interpret the screenshot correctly.

**Comment by mschnitzer** [comment]
@bgeuken it's not possible to display the tooltip also when hovering over the status, cause we already have tooltips what display some different things (like the rpm build log). What do you mean by spacing issue?

File: src/api/app/helpers/webui/buildresult_helper.rb [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+    link_title = status['details']
+    if status['code']
+      code = status['code']
+      theclass = 'status_' + code.gsub(/[- ]/, '_')
+    else
+      code = ''
+      theclass = ' '
+    end
+
+    result = content_tag(:td, class: [theclass, "buildstatus", "nowrap"]) do
```
Comment (ChrisBr) [inline src/api/app/helpers/webui/buildresult_helper.rb]: Nice catch :+1:

**Comment by ChrisBr** [comment]
If you squash your commits it LGTM

**Comment by bgeuken** [comment]
@mschnitzer Good job:+1:

Though next time please split your commits into logical chunks, eg:
* 1 commit for moving the helper code
* 1 commit for doing the html_safe changes
* 1 commit for replacing elide with word_break
etc.

That makes it easier to track what changed. During the review but also later on when bugs appear and one has to find the root cause.

As said, well done:-)
